KIPs/kip-237.md

```diff
@@ -0,0 +1,55 @@
+---
+kip: 237
+title: Storage Compression
+author: Lake (@hyunsooda), Ollie (@blukat29)
+discussions-to: TBD
+status: Draft
+category: Storage
+created: 2024-12-26
+---
+
+## Abstract
+Reduce disk storage by compressing data, cutting its size by 50% over.
+
+## Motivation
+Kaia has focused on optimizing disk storage, effectively reducing the state trie size through state migration and live pruning.
+However, these optimizations only address the state trie data. There is still significant potential for further reductions in storage.
+This KIP targets three types of data: header, body, and receipts.
+
+For each block processed, the header, body, and receipts data are generated.
+These data are progressively stored on disk, consuming a substantial amount of storage.
+Based on the migrated chain data (as of 12/02/2024), the header, body, and receipts account for 10.4%, 28.1%, and 40.9% of the total size, respectively, which together occupy 79.4% of the node's disk space.
+
+We propose implementing storage compression techniques to reduce the size of these three data types—header, body, and receipts—thereby improving disk space utilization.
+
+### Compression
+A node periodically compresses storage in a background thread and removes the original data once it is confirmed safe to do so.
+Data is compressed in units called "chunks," with the chunk size determined by a CLI flag.
+The default chunk size is 1MB. For example, if the target data size is 10MB, the data will be divided into 10 chunks.
+There is a tradeoff between speed and compression rate. Larger chunk sizes result in more effective compression but introduce higher compression and decompression overhead (time).
+
+### Decompression
+Decompression is required when a node receives a query (header, body, or receipts).
+The node first searches for the corresponding data in the original database.
+If the data is not found, it locates the decompressed chunk and performs the decompression. The decompressed chunk is cached for later use.
+
+### Retention
+Retention follows the same semantics as LivePruning, preserving the latest n blocks to prevent them from being compressed.
+
+### Evaluation
+We evaluated compression module on mainnet([chaindata of 12/02/2024](https://packages.kaia.io/mainnet/chaindata/)).
+| Item    | Header | Body | Receipts |
+| ------- | ------- | ------- | ------- |
+| Origin size | 410G | 1100G | 1600G |
+| Compressed size | 261G (40.24%) | 267G (75.73%) | 179G (88.81%) |
+
+Compression had been completed wihtin 3 days.
+
```

Comment:
```suggestion
Evaluation on a [state-migrated mainnet (as of 2024-12-02)](https://packages.kaia.io/mainnet/chaindata/) showed a significant reduction in its disk size. Body and receipts showed high compression rates thanks to their high data repetition.

| Data | Uncompressed (GB) | Compressed (GB) | Reduction (%) |
| Header | 410 | 261 | 36.3% |
| Body | 1100 | 267 | 75.7% |
| Receipts | 1600 | 179 | 88.8% |
| Others | ___ | ___ | - |
| Total | ___ | ___ | __._% |

It took 3 days for background compression threads to complete.
```

I've unified to 3 significant digits.

Reply:
What does `total` means? the sum of reduction for three compression types or total size after compression?

Reply:
Total means the total size of the chaindata directory. I thought readers would be interested about it.

Reply:
Hum.. The total size difference had not been recorded in the sheet.  Unfortunately the chain data is not available due to rotating. I just statically calc. as `4200 - (1160 - 267) - (410 - 261) - (1600 - 179)` and get percentage of it, 57.21%

Reply:
That should do. Thanks!

Reply:
Please fill those numbers in the table. I'm happy with your _static calculation_.

Reply:
Are you okay with leaving Others empty since change is not made by compression?

Reply:
No, I want them on the table. The total chaindata size is very important information. If I would pick 1 piece of information from this whole KIP I'd choose the compression rate of the total chaindata. That's why we need the Others information for completeness.

Let's clarify with "Others (not subject to compression)" or "*Others include state trie and misc database which are not compressed by this KIP".

Reply:
Filed in. Do you mean to change the `Others` field name to another one?

Reply:
I don't have better idea. I think "Others" should do.